File: typescript/date-utils.ts
/**
 * @file Date utilities — pure functions with no temporal side effects.
 */

export function formatDate(date: Date, template = "YYYY-MM-DD"): string {
  const tokens: Record<string, string> = {
    YYYY: date.getFullYear().toString(),
    MM: (date.getMonth() + 1).toString().padStart(2, "0"),
    DD: date.getDate().toString().padStart(2, "0"),
    HH: date.getHours().toString().padStart(2, "0"),
    mm: date.getMinutes().toString().padStart(2, "0"),
    ss: date.getSeconds().toString().padStart(2, "0"),
  };
  return template.replace(/YYYY|MM|DD|HH|mm|ss/g, (k) => tokens[k]);
}

export function daysBetween(a: Date, b: Date): number {
  const ms = Math.abs(a.getTime() - b.getTime());
  return Math.floor(ms / 86_400_000);
}

export function addDays(date: Date, days: number): Date {
  const result = new Date(date);
  result.setDate(result.getDate() + days);
  return result;
}

export function isWeekend(date = new Date()): boolean {
  const day = date.getDay();
  return day === 0 || day === 6;
}

/** Human‑readable elapsed time (e.g. "3h ago"). */
export function relativeTime(date: Date, now = Date.now()): string {
  const seconds = Math.floor((now - date.getTime()) / 1000);
  if (seconds < 60) return "just now";
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `${minutes}m ago`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours}h ago`;
  return `${Math.floor(hours / 24)}d ago`;
}
export function isLeapYear(year: number): boolean {
  return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
}
export function isLeapYear(year: number): boolean {
  return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
}
export function isLeapYear(year: number): boolean {
  return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
}
export function isLeapYear(year: number): boolean {
  return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
}
export function isLeapYear(year: number): boolean {
  return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
}
export function isLeapYear(year: number): boolean {
  return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
}

File: dart/date_utils.dart
/// Date transformation utilities for Dart.
///
/// Pure functions — no side effects, no state.
library tiny_utils.date_utils;

/// Format a [DateTime] as `YYYY-MM-DD`.
String formatDate(DateTime date) => date.toIso8601String().substring(0, 10);

/// Format a [DateTime] as `HH:MM` (24-hour).
String formatTime(DateTime date) =>
    '${date.hour.toString().padLeft(2, '0')}:${date.minute.toString().padLeft(2, '0')}';

/// Return `true` when both dates fall on the same calendar day.
bool isSameDay(DateTime a, DateTime b) =>
    a.year == b.year && a.month == b.month && a.day == b.day;

/// Return `true` when [date] is today.
bool isToday(DateTime date) => isSameDay(date, DateTime.now());

/// Return a human-friendly relative description of a [DateTime].
/// Returns "today", "yesterday", "X days ago", or the formatted date.
String relativeDate(DateTime date) {
  final now = DateTime.now();
  final diff = now.difference(date).inDays;
  if (diff == 0) return 'today';
  if (diff == 1) return 'yesterday';
  if (diff < 7) return '$diff days ago';
  return formatDate(date);
}

/// Return the start of the day (00:00:00.000) for a given [DateTime].
DateTime startOfDay(DateTime date) =>
    DateTime(date.year, date.month, date.day);

/// Return the end of the day (23:59:59.999) for a given [DateTime].
DateTime endOfDay(DateTime date) =>
    DateTime(date.year, date.month, date.day, 23, 59, 59, 999);
